subprojects/core/src/main/java/org/gradle/configuration/DefaultScriptPluginFactory.java

```diff
@@ -178,6 +183,8 @@ public void apply(final Object target) {
             initialRunner.run(target, services);
 
             PluginRequests pluginRequests = initialRunner.getData();
+            pluginRequests = pluginRequestsTransformer.transformPluginRequests(pluginRequests, target);
```

Comment:
:x: This will only apply to a Groovy DSL build script.

The Kotlin DSL provides its own implementation of `ScriptPluginFactory`.
I suppose this transformation of plugin requests should happen in a common place instead.
Probably in `PluginRequestApplicator` right before resolving the requests.

Reply:
If we'd make the transformation call in `PluginRequestApplicator` then we'd also need to pass in the `target` object. In turn that means changing the signature of `PluginRequestApplicator.applyPlugins`. I am guessing that the Kotlin DSL plugin currently relies on a specific signature of that method. 

Wouldn't that be a breaking change? How was that handled in the past? As an alternative you could add the transformation from your implementation of `ScriptPluginFactory` before calling `PluginRequestApplicator`.

Reply:
Right, changing the API of `PluginRequestApplicator` would be a breaking change wrt. Kotlin DSL.

It could be done in three steps, adding a new method and deprecating the other one, then, updating Kotlin DSL, then removing the old method. It's tedious and creates a bunch of follow up work. We did this in the past in similar situations.

In some ways it makes sense that this should be done outside of the applicator whose sole job should be to apply the requests. It should also be done outside of the `PluginRequestCollector` as its sole job is to collect the declared requests, which the Groovy DSL infrastructure caches. In between is the `ScriptPluginFactory`,  that's how things currently are.

So, yes I think it makes sense to use the `PluginRequestTransformer` service in both the Groovy and Kotlin `ScriptPluginFactory` implementations.

WDYT @bamboo?

Reply:
@eskatos As a first step, applying the `PluginRequests` transformation logic on both factories sounds very reasonable to me as long as we keep the duplication to a minimum, see my next comment.